src/attachment.rs

```diff
@@ -0,0 +1,404 @@
+    {
+        this.insert(Cow::Owned(key.to_owned()), Cow::Owned(value.to_owned()));
+    }
+}
+
+/// Associates `value` to `key` in the map, aliasing them.
+///
+/// Note that once `key` is aliased, reinserting at the same key may alias the previous instance, or the new instance of `key`.
+///
+/// Calling this with `NULL` or the gravestone value is undefined behaviour.
+#[no_mangle]
+pub extern "C" fn z_bytes_map_insert_by_alias(
+    this: &z_owned_bytes_map_t,
+    key: z_bytes_t,
+    value: z_bytes_t,
+) {
+    let this = unsafe { &mut *this.get() };
+    if let (Some(this), Some(key), Some(value)) = (this.as_mut(), key.as_slice(), value.as_slice())
+    {
+        unsafe {
+            this.insert(
+                Cow::Borrowed(core::mem::transmute(key)),
+                Cow::Borrowed(core::mem::transmute(value)),
+            )
+        };
+    }
+}
+
+/// Returns the number of key-value pairs in the map.
+///
+/// Calling this with `NULL` or the gravestone value is undefined behaviour.
+#[no_mangle]
+extern "C" fn z_bytes_map_len(this: &z_owned_bytes_map_t) -> usize {
+    let this = unsafe { &*this.get() };
+    this.as_ref().map_or(0, |this| this.len())
+}
+
+/// Iterates over the key-value pairs in the map.
+///
+/// `body` will be called once per pair, with `ctx` as its last argument.
+/// If `body` returns a non-zero value, the iteration will stop immediately and the value will be returned.
+/// Otherwise, this will return 0 once all pairs have been visited.
+/// `body` is not given ownership of the key nor value, which alias the pairs in the map.
+/// It is safe to keep these aliases until existing keys are modified/removed, or the map is destroyed.
+/// Note that this map is unordered.
+///
+/// Calling this with `NULL` or the gravestone value is undefined behaviour.
+#[no_mangle]
+pub extern "C" fn z_bytes_map_iter(
+    this: &z_owned_bytes_map_t,
+    body: z_attachment_iter_body_t,
+    ctx: *mut c_void,
+) -> i8 {
+    let this = unsafe { &*this.get() };
+    if let Some(this) = this.as_ref() {
+        for (key, value) in this.iter() {
+            let result = body(key.as_ref().into(), value.as_ref().into(), ctx);
+            if result != 0 {
+                return result;
+            }
+        }
+    }
+    0
+}
+
+const Z_BYTES_MAP_VTABLE: z_attachment_vtable_t = z_attachment_vtable_t {
+    len: unsafe { core::mem::transmute(z_bytes_map_len as extern "C" fn(_) -> usize) },
+    iteration_driver: unsafe {
+        core::mem::transmute(z_bytes_map_iter as extern "C" fn(_, _, _) -> i8)
+    },
+};
+
+//TODO(sashacmc): avoid to export it to the API, how?
+#[no_mangle]
+pub extern "C" fn insert_in_attachment(key: z_bytes_t, value: z_bytes_t, ctx: *mut c_void) -> i8 {
+    let attachments_ref: &mut Attachment = unsafe { &mut *(ctx as *mut Attachment) };
+    attachments_ref.insert(key.as_slice().unwrap(), value.as_slice().unwrap());
+    0
+}
+
+#[no_mangle]
+extern "C" fn attachment_len(this: *const c_void) -> usize {
```

Comment:
avoid no-mangles for non-public symbols: they still end up in the binary and could cause symbol clashes